File: 279. Perfect Squares/dp.js
/**
 * @param {number} n
 * @return {number}
 */
 var numSquares = function(n) {
  
    const dp = [...Array(n+1)].fill(Infinity);
    dp[0] = 0;
    
    for(let i = 1; i <= n; i++){
        for(let j = 1; j <= Math.floor(Math.sqrt(i)); j++){
            if(i - (j * j) >= 0){
                dp[i] = Math.min(
                    dp[i],
                    dp[i - (j * j)] + 1
                )
            }
        }
    }
    
    return dp[n];
    
};

File: 279. Perfect Squares/dp.py
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [sys.maxsize] * (n + 1);
        dp[0] = 0
        
        max = int(math.sqrt(n)) + 1
        # Doing this in nested arr cost a lot of memory
        square = [i**2 for i in range(1, max)];
        
        for i in range(n+1):
            max = int(math.sqrt(i))
            for sq in square:
                if i - sq < 0:
                    break;
                dp[i] = min(dp[i], dp[i - sq] + 1)
        
        return dp[n]
            
        
                
            
        